### PR TITLE
Failed Tests log by Test

### PR DESCRIPTION
**Pull request by kuznetsovmoci**

### CODE REVIEW

File: common/prettify.pm [hunk 1]
```diff
@@ -471,30 +474,44 @@ sub Subsection ($)
...
+        } 
+        else {
+            $self->{FAILED_TESTS}->{$self->{LAST_SUBSECTION}} = "<h3>$self->{BUILDNAME}</h3>\n$rev_line<br><br>\n";
         }
 
         print {$self->{FH}} "<a name=\"subsection_$self->{SUBSECTION_COUNTER}\"></a>";
         print {$self->{FH}} "<h3>$self->{LAST_SUBSECTION}</h3>\n";
+
+        $self->{CURRENT_SUBSECTION} = $self->{LAST_SUBSECTION};
+
```
Comment (ClaytonCalabrese) [inline common/prettify.pm:512]: Some of these newlines can go away

File: common/prettify.pm [hunk 2]
```diff
@@ -534,10 +557,17 @@ sub Warning ($)
 
         $self->Print_Sections ();
 
-        print {$self->{FH}} "<a name=\"warning_$counter\"></a>\n";
-        print {$self->{FH}} "<tt>[<a href=\"$self->{FULLHTML}#warning_$counter"
-                            . "\">Details</a>] </tt>";
-        print {$self->{FH}} "<font color=\"FF7700\"><tt>$s</tt></font><br>\n";
+        my $Warning1 = "<a name=\"warning_$counter\"></a>\n";
+        my $Warning2 = "<tt>[<a href=\"$self->{FULLHTML}#warning_$counter" . "\">Details</a>] </tt>";
```
Comment (ClaytonCalabrese) [inline common/prettify.pm:561]:
```suggestion
        my $Warning2 = "<tt>[<a href=\"$self->{FULLHTML}#warning_$counter\">Details</a>] </tt>";
```
As these are now on the same line you don't need to append them together and they can be one set of quotes.

File: common/prettify.pm [hunk 3]
```diff
@@ -1099,9 +1131,9 @@ sub new ($$$$$$)
             new Prettify::Totals_HTML ($basename), #Must be 2
             new Prettify::Config_HTML ($basename), #Must be 3
         );
-    
+
     if (!$skip_failed_test_logs) {
-        push @{$self->{OUTPUT}}, new Prettify::Failed_Tests_HTML ($basename, $buildname, $rev_link, $log_prefix); #Must be 4, if used
+        push @{$self->{OUTPUT}}, new Prettify::Failed_Tests_HTML ($basename, $buildname, $self->{FAILED_TESTS}, $rev_link, $log_prefix); #Must be 4, if used
```
Comment (ClaytonCalabrese) [inline common/prettify.pm:1136]: should this comment changed?

Comment (iguessthislldo) [reply common/prettify.pm:1136]: Or removed

File: scoreboard.pl [hunk 4]
```diff
@@ -504,6 +509,9 @@ ($)
 sub update_cache ($)
 {
     my $directory = shift;
+    my %failed_tests_by_test;
+    my $failed_tests_by_test_ref = \%failed_tests_by_test;
+
```
Comment (ClaytonCalabrese) [inline scoreboard.pl:514]:
```suggestion
```

File: scoreboard.pl [hunk 5]
```diff
@@ -543,11 +551,21 @@ ($)
                 }
 
                 print "        Prettifying\n" if($verbose);
-
-                Prettify::Process ("$directory/$buildname/$filename", $buildname, $use_build_logs, $builds{$buildname}->{DIFFROOT}, "$directory/$log_prefix");
+
```
Comment (ClaytonCalabrese) [inline scoreboard.pl:554]:
```suggestion
```

File: scoreboard.pl [hunk 6]
```diff
@@ -572,7 +592,12 @@ ($)
         return;
     }
 
-    my $failed_tests = $directory . "/" . $log_prefix . "_Failed_Tests_By_Build.html";
+    my $failed_tests = $directory  . "/" . $log_prefix .  "_Failed_Tests_By_Build.html";
```
Comment (ClaytonCalabrese) [inline scoreboard.pl:595]:
```suggestion
    my $failed_tests = $directory . "/" . $log_prefix .  "_Failed_Tests_By_Build.html";
```

File: scoreboard.pl [hunk 7]
```diff
@@ -572,7 +592,12 @@ ($)
...
+    if (-e $failed_tests) {
+        unlink $failed_tests;
+    }
+
+    $failed_tests = $directory  . "/" . $log_prefix .  "_Failed_Tests_By_Test.html";
```
Comment (ClaytonCalabrese) [inline scoreboard.pl:600]:
```suggestion
    $failed_tests = $directory . "/" . $log_prefix .  "_Failed_Tests_By_Test.html";
```

File: scoreboard.pl [hunk 8]
```diff
@@ -662,8 +687,8 @@ ($)
         foreach my $file (@existing) {
             if ( -e $file . "_Totals.html" || $post == 1 ) {
                 print "        Prettifying $file.txt\n" if($verbose);
-
-                Prettify::Process ("$file.txt", $buildname, $use_build_logs, $builds{$buildname}->{DIFFROOT}, "$directory/$log_prefix");
+
```
Comment (ClaytonCalabrese) [inline scoreboard.pl:690]:
```suggestion
```

File: scoreboard.pl [hunk 9]
```diff
@@ -789,7 +824,12 @@ ($)
         return;
     }
 
-    my $failed_tests = $directory . "/" . $log_prefix . "_Failed_Tests_By_Build.html";
+    my $failed_tests = $directory  . "/" . $log_prefix .  "_Failed_Tests_By_Build.html";
```
Comment (ClaytonCalabrese) [inline scoreboard.pl:827]:
```suggestion
    my $failed_tests = $directory . "/" . $log_prefix .  "_Failed_Tests_By_Build.html";
```

File: common/prettify.pm (hunk 3 above)
Comment (kuznetsovmoci) [reply common/prettify.pm:1136]: There is logic that relies on this being #4 - similar to other reports above

Comment (ClaytonCalabrese) [reply common/prettify.pm:1135]: Maybe make the comment more clear then. I assumed it meant that it needed 4 variables passed to be used.

**Review by ClaytonCalabrese — CHANGES_REQUESTED** [review]
Only thing I see left to deal with is clarify the comment. Otherwise looks good.

File: common/prettify.pm (hunk 3 above)
Comment (kuznetsovmoci) [reply common/prettify.pm]: I made it more clear, but there is also a context, which makes it even more understandable:

        @{$self->{OUTPUT}} =
            (
                new Prettify::Full_HTML ($basename),   #Must be at 0
                new Prettify::Brief_HTML ($basename),
                new Prettify::Totals_HTML ($basename), #Must be at 2
                new Prettify::Config_HTML ($basename), #Must be at 3
            );

**Review by ClaytonCalabrese — APPROVED** [review]

**Pull request merged**